broadinstitute/single_cell_portal_core
# Migrating all Rails-based tests to use FactoryBot, SelfCleaningSuite (SCP-3061)

Currently, in the Rails-based testing section of SCP, there are a mix of test construction techniques.  Roughly half are older-style tests that rely on data populated at the beginning of each CI run from `db/seeds.rb`.  The other half are newer-style tests that leverage both `FactoryBot` and our internal `SelfCleaningSuite` that handle CRUDing all necessary data for the current suite (including cleanup).  These newer style tests are highly preferable to the older ones as they require no pre-test setup, and developers can run them at any time while working on feature development as they are completely self-contained.

This update migrates all of the older-style tests to this newer framework.  This means that any Rails-based test will now adhere to the following rules:
1. Test suites are now fully idempotent (can be run as many times as needed in any order, and return the same results, regardless of database state)
2. Test suites handle setting their own state, and do not require running `rake db:seed` to populate data
3. Test suites clean up after themselves to remove any extraneous data that might potentially skew downstream results

The main caveat is that some test suites now run slower, as data seeding can sometimes take 10-20s, depending on the kind of data being created (a study w/ Terra workspace takes the longest, and is required for some integration testing).  This can add a non-trivial amount of time to an entire CI run, though the exact amount is not currently known as we lack sufficient data to estimate the impact.  However, given that we intend to eventually parallelize tests as much as possible, this should not be viewed as a blocker.

Additionally, it is still necessary for developers to start `Delayed::Job` and reindex their MongoDB test schema before running tests locally (using the `bin/run_tests.sh` script will handle this for you).

MANUAL TESTING
1. Pull this branch, and then clean out your test schema by running the following commands from the project root:
```
./rails_local_setup.rb && source config/secrets/.source_env.bash
bundle exec bin/rails runner -e test "StudyCleanupTools.destroy_all_studies_and_workspaces"
bundle exec bin/rails runner -e test "BigQueryClient.clear_bq_table"
bundle exec rake RAILS_ENV=test db:purge
bundle exec rake RAILS_ENV=test db:mongoid:create_indexes
```
2. Restart `Delayed::Job` for the test environment:
```
bin/delayed_job restart test -n 6
```
3. Select any Rails-based test suite, and run it 2+ times in a row to ensure idempotency (`SearchControllerTest` and `SiteControllerTest` are good examples):
```
bin/rails test test/api/search_controller_test.rb
bin/rails test test/controllers/site_controller_test.rb
```

This PR satisfies SCP-3061.
